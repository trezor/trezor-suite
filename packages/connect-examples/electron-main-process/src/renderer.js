// print log helper
const printLog = data => {
    const log = document.getElementById('log');
    const current = log.value;
    if (current.length > 0) {
        log.value = `${JSON.stringify(data)}\n\n${current}`;
    } else {
        log.value = JSON.stringify(data);
    }
};

// click to get public key
const btn = document.getElementById('get-xpub');
btn.onclick = () => {
    // send request to trezor-connect
    window.TrezorConnect.send({
        method: 'getPublicKey',
        params: {
            path: "m/49'/0'/0'",
            coin: 'btc',
        },
    });
};

// receive data from TrezorConnect
window.TrezorConnect.receive((event, message) => {
    printLog(message);
});

// init TrezorConnect in electron main process
window.TrezorConnect.init();
