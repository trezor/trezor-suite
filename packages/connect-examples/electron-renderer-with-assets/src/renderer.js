import TrezorConnect, {
    TRANSPORT_EVENT,
    UI,
    UI_EVENT,
    DEVICE_EVENT,
    TRANSPORT,
    DEVICE,
} from '@trezor/connect-web';

// print log helper
const printLog = data => {
    const log = document.getElementById('log');
    const current = log.value;
    if (current.length > 0) {
        log.value = `${JSON.stringify(data)}\n\n${current}`;
    } else {
        log.value = JSON.stringify(data);
    }
};

// SETUP @trezor/connect

// Listen to TRANSPORT_EVENT
TrezorConnect.on(TRANSPORT_EVENT, event => {
    printLog(event.type);
    if (event.type === TRANSPORT.ERROR) {
        // trezor-bridge not installed
        printLog('Transport is missing');
    }
    if (event.type === TRANSPORT.START) {
        printLog(event);
    }
});

// Listen to DEVICE_EVENT
TrezorConnect.on(DEVICE_EVENT, event => {
    printLog(event.type);

    // not obvious event
    if (event.type === DEVICE.CONNECT_UNACQUIRED) {
        // connected device is unknown or busy
        // most common reasons is that either device is currently used somewhere else
        // or app refreshed during call and trezor-bridge didn't managed to release the session
        // render "Acquire device" button and after click try to fetch device features using:
        // TrezorConnect.getFeatures();
    }
});

// Listen to UI_EVENT
// most common requests
TrezorConnect.on(UI_EVENT, event => {
    printLog(event);

    if (event.type === UI.REQUEST_PIN) {
        // example how to respond to pin request
        TrezorConnect.uiResponse({ type: UI.RECEIVE_PIN, payload: '1234' });
    }

    if (event.type === UI.REQUEST_PASSPHRASE) {
        if (event.payload.device.features.capabilities.includes('Capability_PassphraseEntry')) {
            // device does support entering passphrase on device
            // let user choose where to enter
            // if he choose to do it on device respond with:
            TrezorConnect.uiResponse({
                type: UI.RECEIVE_PASSPHRASE,
                payload: { passphraseOnDevice: true, value: '' },
            });
        } else {
            // example how to respond to passphrase request from regular UI input (form)
            TrezorConnect.uiResponse({
                type: UI.RECEIVE_PASSPHRASE,
                payload: { value: 'type your passphrase here', save: true },
            });
        }
    }

    if (event.type === UI.SELECT_DEVICE) {
        if (event.payload.devices.length > 0) {
            // more then one device connected
            // example how to respond to select device
            TrezorConnect.uiResponse({
                type: UI.RECEIVE_DEVICE,
                payload: event.payload.devices[0],
            });
        } else {
            // no devices connected, waiting for connection
        }
    }

    // getAddress from device which is not backed up
    // there is a high risk of coin loss at this point
    // warn user about it
    if (event.type === UI.REQUEST_CONFIRMATION) {
        // payload: true - user decides to continue anyway
        TrezorConnect.uiResponse({ type: UI.RECEIVE_CONFIRMATION, payload: true });
    }
});

console.log('TrezorConnect', TrezorConnect);
// Initialize TrezorConnect
TrezorConnect.init({
    connectSrc: 'trezor-connect-bundled/',
    popup: false, // render your own UI
    webusb: false, // webusb is not supported in electron
    debug: false, // see what's going on inside connect
    // lazyLoad: true, // set to "false" (default) if you want to start communication with bridge on application start (and detect connected device right away)
    // set it to "true", then @trezor/connect will not be initialized until you call some TrezorConnect.method()
    // this is useful when you don't know if you are dealing with Trezor user
    manifest: {
        email: 'email@developer.com',
        appUrl: 'electron-app-boilerplate',
    },
})
    .then(() => {
        printLog('TrezorConnect is ready!');
    })
    .catch(error => {
        printLog('TrezorConnect init error', `TrezorConnect init error:${error}`);
    });

// click to get public key
const btn = document.getElementById('get-xpub');
btn.onclick = () => {
    console.log('click');
    printLog('c');
    TrezorConnect.getPublicKey({
        path: "m/49'/0'/0'",
        coin: 'btc',
    }).then(response => {
        printLog(response);
    });
};
