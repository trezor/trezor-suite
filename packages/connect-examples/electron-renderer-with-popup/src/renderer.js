// TrezorConnect is injected as inline script in html
// therefore it doesn't need to included into node_modules
// get reference straight from window object
const { TrezorConnect } = window;

// print log helper
const printLog = data => {
    const log = document.getElementById('log');
    const current = log.value;
    if (current.length > 0) {
        log.value = `${JSON.stringify(data)}\n\n${current}`;
    } else {
        log.value = JSON.stringify(data);
    }
};

// SETUP @trezor/connect

// Listen to DEVICE_EVENT
// this event will be emitted only after user grants permission to communicate with this app
TrezorConnect.on('DEVICE_EVENT', event => {
    printLog(event);
});

// Initialize TrezorConnect
TrezorConnect.init({
    debug: false, // see whats going on inside iframe
    lazyLoad: true, // set to "false" (default) if you want to start communication with bridge on application start (and detect connected device right away)
    // set it to "true", then @trezor/connect will not be initialized until you call some TrezorConnect.method()
    // this is useful when you don't know if you are dealing with Trezor user
    manifest: {
        email: 'email@developer.com',
        appUrl: 'electron-app-boilerplate',
    },
    transports: ['BridgeTransport'],
})
    .then(() => {
        printLog('TrezorConnect is ready!');
    })
    .catch(error => {
        printLog(`TrezorConnect init error: ${error}`);
    });

// click to get public key
const btn = document.getElementById('get-xpub');
btn.onclick = () => {
    TrezorConnect.getPublicKey({
        path: "m/49'/0'/0'",
        coin: 'btc',
    }).then(response => {
        printLog(response);
    });
};
