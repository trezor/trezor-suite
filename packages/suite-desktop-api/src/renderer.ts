import { DesktopApi } from './api';
import { factory } from './factory';

export const getDesktopApi = () => {
    let api: DesktopApi | undefined;
    if (typeof window !== 'undefined' && process.env.SUITE_TYPE === 'desktop') {
        // it's pointless to write type declaration in global.Window since this is the only reference
        // @ts-expect-error
        api = window.desktopApi;
    }
    return api || factory();
};

export const desktopApi = getDesktopApi();
