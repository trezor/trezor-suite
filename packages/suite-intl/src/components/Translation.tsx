import React from 'react';
import { FormattedMessage, MessageDescriptor } from 'react-intl';
import { messages } from '@suite/messages';

export type TranslationProps = {
    isNested?: boolean;
    isReactNative?: boolean; // used to render span on the web, plain node for react-native
    translationTooltip?: React.ComponentType<TooltipProps>;
} & ExtendedMessageDescriptor;

export interface TooltipProps {
    isNested?: boolean;
    messageId: keyof typeof messages;
    children: any;
}

type PrimitiveType = string | number | boolean | Date | null | undefined;

// Add MessageDescriptor type to values entry
type FormatXMLElementFn = (...args: any[]) => string | object;
export interface ExtendedMessageDescriptor extends MessageDescriptor {
    id: keyof typeof messages;
    values?: {
        [key: string]:
            | PrimitiveType
            | React.ReactElement
            | ExtendedMessageDescriptor
            | FormatXMLElementFn;
    };
}

export const isMsgType = (props: TranslationProps | React.ReactNode): props is TranslationProps =>
    typeof props === 'object' && props !== null && (props as TranslationProps).id !== undefined;

const RenderChildrenComponent = ({ children }: any) => children;

export const Translation = (props: TranslationProps) => {
    // Tooltip is only used in suite-landing-page, in case landing will be moved to eshop this could be removed
    const TooltipComponent = props.translationTooltip ?? RenderChildrenComponent;
    const values: Record<string, PrimitiveType | React.ReactNode | ExtendedMessageDescriptor> = {};
    // message passed via props (id, defaultMessage, values)
    Object.keys(props.values || []).forEach(key => {
        // Iterates through all values. The entry may also contain a MessageDescriptor.
        // If so, Renders MessageDescriptor by passing it to `Translation` component
        const maybeMsg = props.values![key];
        values[key] = isMsgType(maybeMsg) ? (
            <Translation {...maybeMsg} translationTooltip={props.translationTooltip} isNested />
        ) : (
            maybeMsg
        );
    });

    // prevent runtime errors
    if (
        !props.defaultMessage &&
        Object.prototype.hasOwnProperty.call(props, 'id') &&
        !messages[props.id]
    ) {
        return <>{`Unknown translation id: ${props.id}`}</>;
    }

    const defaultTagName = props.isReactNative || props.isNested ? undefined : 'span';

    return (
        <TooltipComponent isNested={props.isNested} messageId={props.id}>
            <FormattedMessage
                id={props.id}
                tagName={defaultTagName}
                defaultMessage={props.defaultMessage || messages[props.id].defaultMessage}
                // pass undefined to a 'values' prop in case of an empty values object
                values={Object.keys(values).length === 0 ? undefined : values}
            />
        </TooltipComponent>
    );
};
