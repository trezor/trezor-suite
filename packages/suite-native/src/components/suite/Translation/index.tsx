import React from 'react';
import BaseTranslation from '@suite/components/suite/Translation/components/BaseTranslation';

type TranslationProps = Omit<React.ComponentProps<typeof BaseTranslation>, 'translationTooltip'>;
const Translation = (props: TranslationProps) => {
    return (
        <BaseTranslation
            {...props}
            isReactNative
            translationTooltip={({ children }) => {
                return children;
            }}
        />
    );
};

export default Translation;
