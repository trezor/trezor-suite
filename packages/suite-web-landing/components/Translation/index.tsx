import React, { useContext } from 'react';
import { Translation as SuiteTranslation } from '@suite/intl';
import { HelperTooltip, HelperTooltipProps } from './HelperTooltip';

export const TranslationModeContext = React.createContext(false);

const CustomHelperTooltip = (props: HelperTooltipProps) => {
    const translationMode = useContext(TranslationModeContext);

    return (
        <HelperTooltip {...props} language="en" translationMode={translationMode}>
            {props.children}
        </HelperTooltip>
    );
};

type TranslationProps = Omit<React.ComponentProps<typeof SuiteTranslation>, 'translationTooltip'>;
const Translation = (props: TranslationProps) => (
    <SuiteTranslation {...props} translationTooltip={CustomHelperTooltip} />
);

export default Translation;
