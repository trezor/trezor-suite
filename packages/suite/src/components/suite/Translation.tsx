import { ReactNode } from 'react';
import { FormattedMessage } from 'react-intl';

import {
    ExtendedMessageDescriptor as CommonExtendedMessageDescriptor,
    FormatXMLElementFn,
} from '@suite-common/intl-types';

// We cannot use aliases here because this file is directly imported by the @suite-common/intl-types
// It's little hacky by this will be solved when PR for refactor intl will be merged.
import messages from '../../support/messages';

export type TranslationKey = keyof typeof messages;

type OwnProps = {
    isNested?: boolean;
};

export type ExtendedMessageDescriptor = CommonExtendedMessageDescriptor;
type MsgType = OwnProps & ExtendedMessageDescriptor;

export const isMsgType = (
    props: MsgType | ReactNode | ExtendedMessageDescriptor | Date | FormatXMLElementFn,
): props is MsgType =>
    typeof props === 'object' && props !== null && (props as MsgType).id !== undefined;

export const Translation = (props: MsgType) => {
    const values: Record<string, any> = {};
    // message passed via props (id, defaultMessage, values)
    Object.keys(props.values || []).forEach(key => {
        // Iterates through all values. The entry may also contain a MessageDescriptor.
        // If so, Renders MessageDescriptor by passing it to `Translation` component
        const maybeMsg = props.values![key];
        values[key] = isMsgType(maybeMsg) ? <Translation {...maybeMsg} isNested /> : maybeMsg;
    });

    // prevent runtime errors
    if (
        !props.defaultMessage &&
        Object.prototype.hasOwnProperty.call(props, 'id') &&
        !messages[props.id]
    ) {
        return <>{`Unknown translation id: ${props.id}`}</>;
    }

    const defaultTagName = props.isNested ? undefined : 'span';

    return (
        <FormattedMessage
            id={props.id}
            tagName={defaultTagName}
            defaultMessage={props.defaultMessage || messages[props.id].defaultMessage}
            // pass undefined to a 'values' prop in case of an empty values object
            values={Object.keys(values).length === 0 ? undefined : values}
        />
    );
};
