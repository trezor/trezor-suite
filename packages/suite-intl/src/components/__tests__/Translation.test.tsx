import React from 'react';

import renderer from 'react-test-renderer';
import { IntlProvider } from 'react-intl';
import { Translation } from '../Translation';
import { Locale } from '../../languages';

const createComponentWithIntl = (children: React.ReactChild, props?: { locale: Locale }) => {
    const { locale = 'en' } = props || {};

    return renderer.create(
        <IntlProvider {...props} locale={locale}>
            {children}
        </IntlProvider>,
    );
};

const messages = {
    TR_HELLO: {
        id: 'TR_HELLO',
        defaultMessage: 'Hello',
        descriptor: 'test desc',
    },
    TR_HELLO_NAME: {
        id: 'TR_HELLO_NAME',
        defaultMessage: 'Hello {TR_NAME}, {TR_AGE}',
    },
    TR_NAME: {
        id: 'TR_NAME',
        defaultMessage: 'Name: {name}',
    },
    TR_AGE: {
        id: 'TR_NAME',
        defaultMessage: 'Age: {age}',
    },
};

describe('Translation component', () => {
    test('with id, defaultMessage props', () => {
        const component = createComponentWithIntl(<Translation id="TR_CANCEL" />);
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('with message that holds string in value (passed via props)', () => {
        const component = createComponentWithIntl(
            // @ts-ignore
            <Translation {...messages.TR_NAME} values={{ name: 'John' }} />,
        );
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('with message that holds another message object in values (passed via props)', () => {
        const component = createComponentWithIntl(
            <Translation
                {...messages.TR_HELLO_NAME}
                values={{
                    // @ts-ignore
                    TR_NAME: { ...messages.TR_NAME, values: { name: 'John' } },
                    TR_AGE: 100,
                }}
            />,
        );
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('with message that holds another in values (passed via props)', () => {
        const component = createComponentWithIntl(
            // @ts-ignore
            <Translation
                {...messages.TR_HELLO_NAME}
                values={{
                    // @ts-ignore
                    TR_NAME: <Translation {...messages.TR_NAME} values={{ name: 'John' }} />,
                    TR_AGE: 100,
                }}
            />,
        );
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});
