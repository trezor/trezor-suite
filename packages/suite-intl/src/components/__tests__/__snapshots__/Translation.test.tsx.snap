// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Translation component with id, defaultMessage props 1`] = `
<span>
  Cancel
</span>
`;

exports[`Translation component with message that holds another in values (passed via props) 1`] = `
<span>
  Hello 
  <span>
    Name: John
  </span>
  , 100
</span>
`;

exports[`Translation component with message that holds another message object in values (passed via props) 1`] = `
<span>
  Hello 
  Name: John
  , 100
</span>
`;

exports[`Translation component with message that holds string in value (passed via props) 1`] = `
<span>
  Name: John
</span>
`;
